README.md
# 0x00DA toolkit  

`ooda` will visit provided addresses and save HTTP code, Title, IP address or redirect.  

## Build  
```
nible build
```

## Usage  

Expected input is one FQDN per line.  

Visit addresses and save to a CSV file  
```
cat example.com.txt | ooda > example.com.csv
```

Generate HTML table from the CSV findings
```
ooda -H -i example.com.csv -o example.com.html
```

## TODO  
```
# lower socket timeout by using tcp.dial directly?
# use async instead of threads?
# use browser user-agent
# sort html results when converting
# diagnoze what kind of ssl error?
# what to do with "parseHtml(body) - IndexDefect" in title?
# sanitize title so we dont get h4cked ourselves
# add -h --help
```


CHANGELOG.md
## 1.0.0 (TBD)

  * Initial release


src/ooda.nim

import
  argparse, strutils, strformat, net, xmltree, httpclient, htmlparser

from streams import newFileStream
from templates import templateHTML, htmlFoot, htmlHeadBody, tr

var
  threads: array[0..4, Thread[seq[string]]] # array of 4 threads
  numThreads = len(threads) # number of threads
  chan: Channel[string]
  ctrlChan: Channel[int] # control channel

let
  csvHeaderRow = "respCode;fqdn;ipAddr;location;title"
  p = newParser("0x00DA"):
    flag("-H", "--html", help = "Produce HTML file")
    option("-o", "--output-file", help = "Output file for the ")
    option("-i", "--input-file", help = "Input file to be converted")
  opts = p.parse()

proc parseHtmlTitle(body: string): string =
  var ret = ""
  try:
    let html = parseHtml(body)
    let titleTag = html.findAll("title")[0]
    let stripped = titleTag.innerText().strip()
    ret = stripped.replace("\n", " ")
  except IndexDefect:
    ret = "n/a [could be second redirect]" # this happen when there is a second redirect
  except AssertionDefect:
    ret = "n/a"
  return ret

proc threadFunc(partialUrls: seq[string]) {.thread.} =
  for url in partialUrls:
    let client = newHttpClient(maxRedirects = 0, timeout = 300)
    var
      msg: string
      ipAddr: string
      location = "n/a"
    # try to connect and collect basic information
    try:
      let response = client.request(url, httpMethod = HttpGet)
      # get the IP address
      ipAddr = client.getSocket.getPeerAddr[0]
      # close the connection
      client.close()
      let title = parseHtmlTitle(response.body)
      # if there is a redirect, collect the location
      if response.status.startsWith("30"):
        location = response.headers.getOrDefault("location")
      msg = fmt"{response.status[.. 3]};{url};{ipAddr};{location};{title}"
    except SslError:
      continue
    except OSError:  # Connection timed out
      continue
    except ProtocolError:  # Connection was closed
      continue
    except TimeoutError:  # Timed out
      continue
    # sent the results to chan
    chan.send(msg)
  # exit thread
  ctrlChan.send(1)

when isMainModule:
  if opts.html:
    # Convert CSV file to a HTML table
    if opts.output_file == "" or opts.input_file == "":
      echo "Define input/output files to be converted."
    else:
      echo "Generating HTML table file..."
      templateHTML(opts.input_file, opts.output_file)
  else:
    # all urls loaded from stdin
    var urls: seq[string]
    # Parse stdin input line by line and try to observe where it leads to
    while not endOfFile(stdin):
      var url: string = readLine(stdin)
      # if no uri scheme is supplied, use https
      if not url.startsWith("http://") or url.startsWith("https://"):
        url = "https://" & url
        urls.add(url)
    # divide urls to be crawled by multiple threads
    var inputLen = len(urls)
    var fairShare = int (inputLen / numThreads)
    var firstShare = inputLen - ((numThreads - 1) * fairShare)
    # print the CSV header first
    echo csvHeaderRow
    # create channels for control and crawler
    ctrlChan.open()
    chan.open()
    # create the crawler threads and provide them urls
    # first thread
    var index = firstShare
    createThread(threads[0], threadFunc, urls[0..firstShare - 1])
    # the rest of threads
    for i in 1..high(threads):
      createThread(threads[i], threadFunc, urls[index..index + fairShare - 1])
      # calculate index for the next thread
      index = index + fairShare
    var finished = 0
    while true:
      # receive the actual result in a non-blocking way
      let crawlerMsg = chan.tryRecv()
      if crawlerMsg.dataAvailable:
        echo crawlerMsg.msg
      # receive control message once the thread ended
      # This must run after we receive the actual messages !
      let ctrlMsg = ctrlChan.tryRecv()
      if ctrlMsg.dataAvailable:
        finished = finished + 1
        if finished >= numThreads:
          # all threads finished
          break
      sleep(10)
    joinThreads(threads)
    chan.close()
    ctrlChan.close()


src/templates.nim
# Defined HTML templates and the support functions

import
  strutils, parsecsv, tables

var
  pcsv: CsvParser
let
  colors = {"200": "okColor", "300": "redirColor", "400": "notfoundColor"}.toTable
  htmlFoot* = """</tbody></table></div></body></html>"""
  tr* = """
    <tr>
        <td><span class="%%color">%%respCode</span></td>
        <td><a href="%%fqdn">%%fqdn</a></td>
        <td>%%title</td>
        <td><span class="location">%%location</span></td>
        <td>%%ipAddr</td>
    </tr>
    """
  htmlHeadBody* = """
      <!DOCTYPE html><html><head><title>0x00DA</title>
      <style>
          .okColor { background-color: rgba(60, 179, 113, .4); }
          .redirColor { background-color: rgba(255, 165, 0, .4); }
          .notfoundColor { background-color: rgba(255, 99, 71, .4); }
          .unknownRespCode { background-color: rgba(180, 180, 180, .4); }
      </style>
      </head>
      <script>
        /*!
         * tsorter 2.0.0 - Copyright 2015 Terrill Dent, http://terrill.ca
         * JavaScript HTML Table Sorter
         * Released under MIT license, http://terrill.ca/sorting/tsorter/LICENSE
         */
        var tsorter=function(){"use strict";var a,b,c,d=!!document.addEventListener;return Object.create||(Object.create=function(a){var b=function(){return void 0};return b.prototype=a,new b}),b=function(a,b,c){d?a.addEventListener(b,c,!1):a.attachEvent("on"+b,c)},c=function(a,b,c){d?a.removeEventListener(b,c,!1):a.detachEvent("on"+b,c)},a={getCell:function(a){var b=this;return b.trs[a].cells[b.column]},sort:function(a){var         b=this,c=a.target;b.column=c.cellIndex,b.get=b.getAccessor(c.getAttribute("data-tsorter")),b.prevCol===b.column?(c.className="ascend"!==c.className?"ascend":"descend",b.reverseTable()):(c.className="ascend",-1!==b.prevCol&&"exc_cell"!==b.ths[b.prevCol].className&&(b.ths[b.prevCol].className=""),b.quicksort(0,b.trs.length)),b.prevCol=b.column},getAccessor:function(a){var b=this,c=b.accessors;if(c&&c[a])return c[a];switch(a){case"link":return function(a){return         b.getCell(a).firstChild.firstChild.nodeValue};case"input":return function(a){return b.getCell(a).firstChild.value};case"numeric":return function(a){return parseFloat(b.getCell(a).firstChild.nodeValue,10)};default:return function(a){return b.getCell(a).firstChild.nodeValue}}},exchange:function(a,b){var         c,d=this,e=d.tbody,f=d.trs;a===b+1?e.insertBefore(f[a],f[b]):b===a+1?e.insertBefore(f[b],f[a]):(c=e.replaceChild(f[a],f[b]),f[a]?e.insertBefore(c,f[a]):e.appendChild(c))},reverseTable:function(){var a,b=this;for(a=1;a<b.trs.length;a++)b.tbody.insertBefore(b.trs[a],b.trs[0])},quicksort:function(a,b){var c,d,e,f=this;if(!(a+1>=b)){if(b-a===2)return void(f.get(b-1)>f.get(a)&&f.exchange(b-1,a));for(c=a+1,d=b-        1,f.get(a)>f.get(c)&&f.exchange(c,a),f.get(d)>f.get(a)&&f.exchange(a,d),f.get(a)>f.get(c)&&f.exchange(c,a),e=f.get(a);;){for(d--;e>f.get(d);)d--;for(c++;f.get(c)>e;)c++;if(c>=d)break;f.exchange(c,d)}f.exchange(a,d),b-d>d-a?(f.quicksort(a,d),f.quicksort(d+1,b)):(f.quicksort(d+1,b),f.quicksort(a,d))}},init:function(a,c,d){var e,f=this;for("string"==typeof         a&&(a=document.getElementById(a)),f.table=a,f.ths=a.getElementsByTagName("th"),f.tbody=a.tBodies[0],f.trs=f.tbody.getElementsByTagName("tr"),f.prevCol=c&&c>0?c:-1,f.accessors=d,f.boundSort=f.sort.bind(f),e=0;e<f.ths.length;e++)b(f.ths[e],"click",f.boundSort)},destroy:function(){var a,b=this;if(b.ths)for(a=0;a<b.ths.length;a++)c(b.ths[a],"click",b.boundSort)}},{create:function(b,c,d){var e=Object.create(a);return e.init(b,c,d),e}}}();
        function init() {
            // custom sorter functions
            var sorter = tsorter.create('results', 0, {
                "http-code": function(row){
                  return parseInt( this.getCell(row).childNodes[0].textContent);
                },
                "location": function(row){
                  console.log(this.getCell(row).childNodes[0].textContent);
                  return this.getCell(row).childNodes[0].textContent;
                }
            });
            // click for initial sort
            document.getElementsByTagName("th")[0].click();
        }
        window.onload = init;
      </script>
      <body>
      <div class="container">
      <table id="results">
      <thead>
        <tr>
          <th data-tsorter="http-code">code</th>
          <th>fqdn</th>
          <th>title</th>
          <th>location</th>
          <th>ip address</th>
        </tr>
     </thead>
     <tbody>
      """

proc templateHTML *(inFile: string, outFile: string) =
  var col = ""
  let fw = open(outFile, fmWrite)
  # Write HTML header
  fw.write(htmlHeadBody)
  pcsv.open(infile, separator = ';')
  pcsv.readHeaderRow()
  while pcsv.readRow(columns = 5):
    var color: string
    if pcsv.rowEntry("respCode").startswith("2"): color = colors["200"]
    elif pcsv.rowEntry("respCode").startswith("3"): color = colors["300"]
    elif pcsv.rowEntry("respCode").startswith("4"): color = colors["400"]
    else: color = "unknownRespCode"
    col = tr.multiReplace(
      ("%%color", color),
      ("%%respCode", pcsv.rowEntry("respCode").strip()),
      ("%%fqdn", pcsv.rowEntry("fqdn").strip()),
      ("%%ipAddr", pcsv.rowEntry("ipAddr").strip()),
      ("%%location", pcsv.rowEntry("location").strip()),
      ("%%title", pcsv.rowEntry("title").strip())
    )
    fw.write(col)
  close(pcsv)
  fw.write(htmlFoot)
  close(fw)
